apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc.rs

```diff
@@ -0,0 +1,245 @@
+        }
+    }
+}
+
+impl<State> BlockSpaceAllocator<State> {
+    /// Return the amount of space left to initialize in all
+    /// [`TxBin`] instances.
+    ///
+    /// This is calculated based on the difference between the Tendermint
+    /// block space for a given round and the sum of the allotted space
+    /// to each [`TxBin`] instance in a [`BlockSpaceAllocator`].
+    #[inline]
+    fn uninitialized_space_in_bytes(&self) -> u64 {
+        let total_bin_space = self.protocol_txs.allotted_space_in_bytes
+            + self.encrypted_txs.allotted_space_in_bytes
+            + self.decrypted_txs.allotted_space_in_bytes;
+        self.block.allotted_space_in_bytes - total_bin_space
+    }
+
+    /// Claim all the space used by the [`TxBin`] instances
+    /// as block space.
+    #[inline]
+    fn claim_block_space(&mut self) {
+        let used_space = self.protocol_txs.occupied_space_in_bytes
+            + self.encrypted_txs.occupied_space_in_bytes
+            + self.decrypted_txs.occupied_space_in_bytes;
+
+        self.block.occupied_space_in_bytes = used_space;
+
+        self.decrypted_txs = TxBin::default();
+        self.protocol_txs = TxBin::default();
+        self.encrypted_txs = TxBin::default();
+    }
+}
+
+/// Allotted space for a batch of transactions of the same kind in some
+/// proposed block, measured in bytes.
+#[derive(Debug, Copy, Clone, Default)]
+struct TxBin {
+    /// The current space utilized by the batch of transactions.
+    occupied_space_in_bytes: u64,
+    /// The maximum space the batch of transactions may occupy.
+    allotted_space_in_bytes: u64,
+}
+
+impl TxBin {
+    /// Return a new [`TxBin`] with a total allotted space equal to the
+    /// floor of the fraction `frac` of the available block space `max_bytes`.
+    #[inline]
+    fn init_over_ratio(max_bytes: u64, frac: threshold::Threshold) -> Self {
+        let allotted_space_in_bytes = frac.over(max_bytes);
+        Self {
+            allotted_space_in_bytes,
+            occupied_space_in_bytes: 0,
+        }
+    }
+
+    /// Return the amount of space left in this [`TxBin`].
+    #[inline]
+    fn space_left_in_bytes(&self) -> u64 {
+        self.allotted_space_in_bytes - self.occupied_space_in_bytes
+    }
+
+    /// Construct a new [`TxBin`], with a capacity of `max_bytes`.
+    #[inline]
+    fn init(max_bytes: u64) -> Self {
+        Self {
+            allotted_space_in_bytes: max_bytes,
+            occupied_space_in_bytes: 0,
+        }
+    }
+
+    /// Shrink the allotted space of this [`TxBin`] to whatever
+    /// space is currently being utilized.
+    #[inline]
+    fn shrink_to_fit(&mut self) {
+        self.allotted_space_in_bytes = self.occupied_space_in_bytes;
+    }
+
+    /// Try to dump a new transaction into this [`TxBin`].
+    ///
+    /// Signal the caller if the tx is larger than its max
+    /// allotted bin space.
+    fn try_dump<'tx>(&mut self, tx: &'tx [u8]) -> AllocStatus<'tx> {
```

Comment:
We don't need to return the `tx` field in `AllocStatus` anymore, we'd just be passing back to the caller what they gave us

Reply:
Is dumping the same as allocating a tx? Otherwise could this be called `try_alloc` instead of `try_dump` (especially since it returns `AllocStatus`)? Or will this eventually put the transaction bytes somewhere?